internal/service/kendra/index.go

```diff
@@ -0,0 +1,1036 @@
+		if d.HasChange("user_token_configurations") {
+			input.UserTokenConfigurations = expandUserTokenConfigurations(d.Get("user_token_configurations").([]interface{}))
+		}
+
+		_, err := conn.UpdateIndex(ctx, input)
+
+		if err != nil {
+			return diag.Errorf("[ERROR] Error updating Index (%s): %s", d.Id(), err)
+		}
+
+		// waiter since the status changes from UPDATING to either ACTIVE or FAILED
+		if _, err := waitIndexUpdated(ctx, conn, d.Id(), d.Timeout(schema.TimeoutUpdate)); err != nil {
+			return diag.Errorf("error waiting for Index (%s) update: %s", d.Id(), err)
+		}
+	}
+
+	if d.HasChange("tags_all") {
+		o, n := d.GetChange("tags_all")
+		if err := UpdateTags(ctx, conn, d.Get("arn").(string), o, n); err != nil {
+			return diag.Errorf("error updating tags: %s", err)
+		}
+	}
+
+	return resourceIndexRead(ctx, d, meta)
+}
+
+func resourceIndexDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).KendraConn
+
+	id := d.Id()
+
+	_, err := conn.DeleteIndex(ctx, &kendra.DeleteIndexInput{
+		Id: aws.String(id),
+	})
+
+	if err != nil {
+		return diag.Errorf("error deleting Index (%s): %s", d.Id(), err)
+	}
+
+	if _, err := waitIndexDeleted(ctx, conn, id, d.Timeout(schema.TimeoutDelete)); err != nil {
+		return diag.Errorf("error waiting for Index (%s) delete: %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func findIndexByID(ctx context.Context, conn *kendra.Client, id string) (*kendra.DescribeIndexOutput, error) {
+	input := &kendra.DescribeIndexInput{
+		Id: aws.String(id),
+	}
+
+	output, err := conn.DescribeIndex(ctx, input)
+
+	if err != nil {
+		var resourceNotFoundException *types.ResourceNotFoundException
+
+		if errors.As(err, &resourceNotFoundException) {
+			return nil, &resource.NotFoundError{
+				LastError:   err,
+				LastRequest: input,
+			}
+		}
+
+		return nil, err
+	}
+
+	if output == nil {
+		return nil, tfresource.NewEmptyResultError(input)
+	}
+
+	return output, nil
+}
+
+func statusIndex(ctx context.Context, conn *kendra.Client, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		output, err := findIndexByID(ctx, conn, id)
+
+		if tfresource.NotFound(err) {
+			return nil, "", nil
+		}
+
+		if err != nil {
+			return nil, "", err
+		}
+
+		return output, string(output.Status), nil
+	}
+}
+
+func waitIndexCreated(ctx context.Context, conn *kendra.Client, id string, timeout time.Duration) (*kendra.DescribeIndexOutput, error) {
+
+	stateConf := &resource.StateChangeConf{
+		Pending: IndexStatusValues(types.IndexStatusCreating),
+		Target:  IndexStatusValues(types.IndexStatusActive, types.IndexStatusFailed),
+		Timeout: timeout,
+		Refresh: statusIndex(ctx, conn, id),
+	}
+
+	outputRaw, err := stateConf.WaitForStateContext(ctx)
+
+	if output, ok := outputRaw.(*kendra.DescribeIndexOutput); ok {
+		tfresource.SetLastError(err, errors.New(aws.ToString(output.ErrorMessage)))
+
+		return output, err
+	}
+
+	return nil, err
+}
```

Comment:
to catch the failed state error, I think we should move it out of the target state slice and catch it like the following:
```suggestion
func waitIndexCreated(ctx context.Context, conn *kendra.Client, id string, timeout time.Duration) (*kendra.DescribeIndexOutput, error) {

	stateConf := &resource.StateChangeConf{
		Pending: IndexStatusValues(types.IndexStatusCreating),
		Target:  IndexStatusValues(types.IndexStatusActive),
		Timeout: timeout,
		Refresh: statusIndex(ctx, conn, id),
	}

	outputRaw, err := stateConf.WaitForStateContext(ctx)

	if output, ok := outputRaw.(*kendra.DescribeIndexOutput); ok {
	        if output.Status == types.IndexStatusFailed {
			tfresource.SetLastError(err, errors.New(aws.ToString(output.ErrorMessage)))
		}

		return output, err
	}

	return nil, err
}
```